runtime/lua/vim/diagnostic.lua

```diff
@@ -1440,9 +1440,11 @@ M.handlers.signs = {
 
     local numhl = opts.signs.numhl or {}
     local linehl = opts.signs.linehl or {}
+    local buf_loaded = api.nvim_buf_is_loaded(bufnr)
+    local line_count = buf_loaded and api.nvim_buf_line_count(bufnr) or 0
 
     for _, diagnostic in ipairs(diagnostics) do
-      if api.nvim_buf_is_loaded(diagnostic.bufnr) then
+      if buf_loaded and diagnostic.lnum <= line_count then
```

Comment:
No need to iterate over each diagnostic if the bufnr is bad. Return early way higher up in the function.